lib/data/external/AzureClient.js

```diff
@@ -125,6 +236,74 @@ class AzureClient {
         });
     }
 
+    uploadPart(request, streamingV4Params, partStream, size, key, uploadId,
+    partNumber, bucket, log, callback) {
+        const azureKey = this._createAzureKey(bucket, key, this._bucketMatch);
+        const contentMD5 = request.headers['content-md5'];
+
+        const params = { bucketName: this._azureContainerName,
+            partNumber, size, objectKey: azureKey, contentMD5, uploadId };
+        const dataRetrievalInfo = {
+            key: partNumber,
+            dataStoreName: this._dataStoreName,
+            dataStoreType: 'azure',
+        };
+
+        if (size === 0) {
+            // Azure does not allow putting empty blocks; instead we
+            // will just upload a summary part with content-length of 0.
+            log.debug('0-byte part does not store data',
+                { method: 'uploadPart' });
+            dataRetrievalInfo.dataStoreETag = azureMpuUtils.zeroByteETag;
+            dataRetrievalInfo.numberSubParts = 0;
+            return callback(null, dataRetrievalInfo);
+        }
+        if (size <= azureMpuUtils.maxSubPartSize) {
+            return this._putSinglePart(request, params, log,
+            (err, dataStoreETag, numberSubParts) => {
+                if (err) {
+                    return callback(err);
+                }
+                dataRetrievalInfo.dataStoreETag = dataStoreETag;
+                dataRetrievalInfo.numberSubParts = numberSubParts;
+                return callback(null, dataRetrievalInfo);
+            });
+        }
+        return this._putSubParts(request, params, log,
+        (err, dataStoreETag, numberSubParts) => {
+            if (err) {
+                callback(err);
+            }
+            dataRetrievalInfo.dataStoreETag = dataStoreETag;
+            dataRetrievalInfo.numberSubParts = numberSubParts;
+            return callback(null, dataRetrievalInfo);
+        });
+    }
+
+    abortMPU(key, uploadId, bucket, log, callback) {
+        const azureBucket = this._azureContainerName;
+        const azureKey = this._createAzureKey(bucket, key, this._bucketMatch);
+        return this._client.deleteBlob(azureBucket, azureKey, err => {
```

Comment:
what happens if there was already a fully fledged key under this name and blocks were added and then there is an abort? assuming you can add blocks to an existing key, we don't want to wipe out the original key on the abort call.